File: inference_profiler/collectors/cpu.py
import glob
import os

from .base import BaseCollector


class CpuCollector(BaseCollector):
    @staticmethod
    def collect():
        stat_metrics, t_stat = CpuCollector._get_proc_stat()
        load_avg, t_load = CpuCollector._get_load_avg()
        cpu_mhz, t_freq = CpuCollector._get_cpu_freq()

        metrics = {
            # CPU time metrics (in centiseconds, 1 jiffy = 1/100 sec)
            "vCpuTimeUserMode": stat_metrics['user'] * BaseCollector.JIFFIES_PER_SECOND,
            "tvCpuTimeUserMode": t_stat,
            "vCpuTimeKernelMode": stat_metrics['system'] * BaseCollector.JIFFIES_PER_SECOND,
            "tvCpuTimeKernelMode": t_stat,
            "vCpuIdleTime": stat_metrics['idle'] * BaseCollector.JIFFIES_PER_SECOND,
            "tvCpuIdleTime": t_stat,
            "vCpuTimeIOWait": stat_metrics['iowait'] * BaseCollector.JIFFIES_PER_SECOND,
            "tvCpuTimeIOWait": t_stat,
            "vCpuTimeIntSrvc": stat_metrics['irq'] * BaseCollector.JIFFIES_PER_SECOND,
            "tvCpuTimeIntSrvc": t_stat,
            "vCpuTimeSoftIntSrvc": stat_metrics['softirq'] * BaseCollector.JIFFIES_PER_SECOND,
            "tvCpuTimeSoftIntSrvc": t_stat,
            "vCpuNice": stat_metrics['nice'] * BaseCollector.JIFFIES_PER_SECOND,
            "tvCpuNice": t_stat,
            "vCpuSteal": stat_metrics['steal'] * BaseCollector.JIFFIES_PER_SECOND,
            "tvCpuSteal": t_stat,
            # Total CPU time (user + kernel) in centiseconds
            "vCpuTime": (stat_metrics['user'] + stat_metrics['system']) * BaseCollector.JIFFIES_PER_SECOND,
            "tvCpuTime": t_stat,
            # Context switches
            "vCpuContextSwitches": stat_metrics['ctxt'],
            "tvCpuContextSwitches": t_stat,
            # Load average and frequency
            "vLoadAvg": load_avg,
            "tvLoadAvg": t_load,
            "vCpuMhz": cpu_mhz,
            "tvCpuMhz": t_freq,
        }
        return metrics

    @staticmethod
    def _get_proc_stat():
        metrics = {
            'user': 0, 'nice': 0, 'system': 0, 'idle': 0,
            'iowait': 0, 'irq': 0, 'softirq': 0, 'steal': 0, 'ctxt': 0
        }
        lines, timestamp = BaseCollector._get_file_lines('/proc/stat')

        for line in lines:
            parts = line.split()
            if not parts:
                continue

            if parts[0] == 'cpu':
                metrics['user'] = int(parts[1])
                metrics['nice'] = int(parts[2])
                metrics['system'] = int(parts[3])
                metrics['idle'] = int(parts[4])
                if len(parts) > 5:
                    metrics['iowait'] = int(parts[5])
                if len(parts) > 6:
                    metrics['irq'] = int(parts[6])
                if len(parts) > 7:
                    metrics['softirq'] = int(parts[7])
                if len(parts) > 8:
                    metrics['steal'] = int(parts[8])
            elif parts[0] == 'ctxt':
                metrics['ctxt'] = int(parts[1])

        return metrics, timestamp

    @staticmethod
    def _get_load_avg():
        content, timestamp = BaseCollector._probe_file('/proc/loadavg')
        try:
            return float(content.split()[0]), timestamp
        except (AttributeError, ValueError, IndexError):
            return 0.0, timestamp

    @staticmethod
    def _get_cpu_freq():
        timestamp = BaseCollector.get_timestamp()
        freq_mhz = 0.0

        # Method 1: SysFS
        freqs = []
        files = glob.glob('/sys/devices/system/cpu/cpu*/cpufreq/scaling_cur_freq')
        for p in files:
            val, _ = BaseCollector._read_int(p)
            if val > 0:
                freqs.append(val)

        if freqs:
            freq_mhz = (sum(freqs) / len(freqs)) / 1000.0
            return freq_mhz, timestamp

        # Method 2: /proc/cpuinfo fallback
        lines, _ = BaseCollector._get_file_lines('/proc/cpuinfo')
        for line in lines:
            if line.startswith('cpu MHz'):
                try:
                    freq_mhz = float(line.split(':')[1])
                    break
                except ValueError:
                    pass

        return freq_mhz, timestamp

    @staticmethod
    def get_static_info():
        return {
            "vNumProcessors": os.cpu_count() or 0,
            "vCpuType": CpuCollector._get_cpu_type(),
            "vCpuCache": CpuCollector._get_cpu_cache(),
            "vKernelInfo": CpuCollector._get_kernel_info(),
        }

    @staticmethod
    def _get_kernel_info():
        try:
            u = os.uname()
            return f"{u.sysname} {u.nodename} {u.release} {u.version} {u.machine}"
        except Exception:
            return "unknown"

    @staticmethod
    def _get_cpu_type():
        lines, _ = BaseCollector._get_file_lines('/proc/cpuinfo')
        for line in lines:
            if line.startswith('model name'):
                return line.split(':', 1)[1].strip()
        return "unknown"

    @staticmethod
    def _get_cpu_cache():
        cache_map = {}
        for index_dir in glob.glob('/sys/devices/system/cpu/cpu*/cache/index*'):
            level, _ = BaseCollector._probe_file(os.path.join(index_dir, "level"))
            cache_type, _ = BaseCollector._probe_file(os.path.join(index_dir, "type"))
            size_str, _ = BaseCollector._probe_file(os.path.join(index_dir, "size"))
            shared_cpu_map, _ = BaseCollector._probe_file(os.path.join(index_dir, "shared_cpu_map"))

            if not (level and size_str):
                continue

            suffix = ""
            if cache_type == "Data":
                suffix = "d"
            elif cache_type == "Instruction":
                suffix = "i"

            key = f"L{level}{suffix}"

            # Convert size to bytes
            multiplier = 1
            if size_str.endswith('K'):
                multiplier = 1024
                size_str = size_str[:-1]
            elif size_str.endswith('M'):
                multiplier = 1024 * 1024
                size_str = size_str[:-1]

            try:
                size_bytes = int(size_str) * multiplier
            except ValueError:
                continue

            if key not in cache_map:
                cache_map[key] = {}
            # Deduplicate based on shared cpu map
            if shared_cpu_map not in cache_map[key]:
                cache_map[key][shared_cpu_map] = size_bytes

        result = {}
        for key, cpu_maps in cache_map.items():
            result[key] = sum(cpu_maps.values())

        return result

File: src/cpu.go
package src

import (
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

// CPUMetrics contains VM-level CPU measurements.
// All time values are in centiseconds (1 cs = 10ms).
type CPUMetrics struct {
	// Aggregate CPU time
	Time       exporter.Timed[int64] `json:"vCpuTime"`            // User + Kernel time
	UserMode   exporter.Timed[int64] `json:"vCpuTimeUserMode"`    // User space execution
	KernelMode exporter.Timed[int64] `json:"vCpuTimeKernelMode"`  // Kernel execution
	Idle       exporter.Timed[int64] `json:"vCpuIdleTime"`        // Idle state
	IOWait     exporter.Timed[int64] `json:"vCpuTimeIOWait"`      // Waiting for I/O
	IRQ        exporter.Timed[int64] `json:"vCpuTimeIntSrvc"`     // Hardware interrupts
	SoftIRQ    exporter.Timed[int64] `json:"vCpuTimeSoftIntSrvc"` // Software interrupts
	Nice       exporter.Timed[int64] `json:"vCpuNice"`            // Low priority user procs
	Steal      exporter.Timed[int64] `json:"vCpuSteal"`           // Stolen by hypervisor

	// Performance counters
	ContextSwitches exporter.Timed[int64]   `json:"vCpuContextSwitches"` // Total context switches
	LoadAvg         exporter.Timed[float64] `json:"vLoadAvg"`            // 1-min load average
	FreqMHz         exporter.Timed[float64] `json:"vCpuMhz"`             // Current avg frequency
}

// CPUStatic contains static CPU information.
type CPUStatic struct {
	NumProcessors int              `json:"vNumProcessors"` // Logical CPU count
	Model         string           `json:"vCpuType"`       // Processor model
	Cache         map[string]int64 `json:"vCpuCache"`      // Cache sizes (L1d, L1i, L2, L3)
	KernelInfo    string           `json:"vKernelInfo"`    // Kernel version
}

const jiffiesPerSec = 100 // Standard Linux jiffy rate

// CollectCPU gathers CPU metrics from /proc/stat, /proc/loadavg, and sysfs.
func CollectCPU() CPUMetrics {
	stat, ts := parseProcStat()
	load, loadTS := parseLoadAvg()
	freq, freqTS := getCPUFreq()

	user := stat["user"]
	system := stat["system"]

	return CPUMetrics{
		Time:            exporter.TimedAt((user+system)*jiffiesPerSec, ts),
		UserMode:        exporter.TimedAt(user*jiffiesPerSec, ts),
		KernelMode:      exporter.TimedAt(system*jiffiesPerSec, ts),
		Idle:            exporter.TimedAt(stat["idle"]*jiffiesPerSec, ts),
		IOWait:          exporter.TimedAt(stat["iowait"]*jiffiesPerSec, ts),
		IRQ:             exporter.TimedAt(stat["irq"]*jiffiesPerSec, ts),
		SoftIRQ:         exporter.TimedAt(stat["softirq"]*jiffiesPerSec, ts),
		Nice:            exporter.TimedAt(stat["nice"]*jiffiesPerSec, ts),
		Steal:           exporter.TimedAt(stat["steal"]*jiffiesPerSec, ts),
		ContextSwitches: exporter.TimedAt(stat["ctxt"], ts),
		LoadAvg:         exporter.TimedAt(load, loadTS),
		FreqMHz:         exporter.TimedAt(freq, freqTS),
	}
}

// CollectCPUStatic gathers static CPU info.
func CollectCPUStatic() CPUStatic {
	return CPUStatic{
		NumProcessors: getNumCPU(),
		Model:         getCPUModel(),
		Cache:         getCPUCache(),
		KernelInfo:    getKernelInfo(),
	}
}

// parseProcStat parses /proc/stat for CPU and context switch counters.
func parseProcStat() (map[string]int64, int64) {
	lines, ts := exporter.readLines("/proc/stat")
	m := map[string]int64{
		"user": 0, "nice": 0, "system": 0, "idle": 0,
		"iowait": 0, "irq": 0, "softirq": 0, "steal": 0, "ctxt": 0,
	}

	for _, line := range lines {
		fields := strings.Fields(line)
		if len(fields) == 0 {
			continue
		}

		switch fields[0] {
		case "cpu":
			// cpu user nice system idle iowait irq softirq steal guest guest_nice
			if len(fields) > 1 {
				m["user"], _ = strconv.ParseInt(fields[1], 10, 64)
			}
			if len(fields) > 2 {
				m["nice"], _ = strconv.ParseInt(fields[2], 10, 64)
			}
			if len(fields) > 3 {
				m["system"], _ = strconv.ParseInt(fields[3], 10, 64)
			}
			if len(fields) > 4 {
				m["idle"], _ = strconv.ParseInt(fields[4], 10, 64)
			}
			if len(fields) > 5 {
				m["iowait"], _ = strconv.ParseInt(fields[5], 10, 64)
			}
			if len(fields) > 6 {
				m["irq"], _ = strconv.ParseInt(fields[6], 10, 64)
			}
			if len(fields) > 7 {
				m["softirq"], _ = strconv.ParseInt(fields[7], 10, 64)
			}
			if len(fields) > 8 {
				m["steal"], _ = strconv.ParseInt(fields[8], 10, 64)
			}
		case "ctxt":
			if len(fields) > 1 {
				m["ctxt"], _ = strconv.ParseInt(fields[1], 10, 64)
			}
		}
	}
	return m, ts
}

// parseLoadAvg reads the 1-minute load average from /proc/loadavg.
func parseLoadAvg() (float64, int64) {
	content, ts := exporter.readFile("/proc/loadavg")
	fields := strings.Fields(content)
	if len(fields) == 0 {
		return 0, ts
	}
	v, _ := strconv.ParseFloat(fields[0], 64)
	return v, ts
}

// getCPUFreq returns average CPU frequency in MHz.
// Tries sysfs first, falls back to /proc/cpuinfo.
func getCPUFreq() (float64, int64) {
	ts := nowMilli()

	// Try sysfs scaling_cur_freq (in kHz)
	pattern := "/sys/devices/system/cpu/cpu*/cpufreq/scaling_cur_freq"
	matches, _ := filepath.Glob(pattern)

	if len(matches) > 0 {
		var sum int64
		var count int
		for _, p := range matches {
			v, _ := exporter.readInt(p)
			if v > 0 {
				sum += v
				count++
			}
		}
		if count > 0 {
			return float64(sum) / float64(count) / 1000.0, ts
		}
	}

	// Fallback to /proc/cpuinfo
	lines, ts := exporter.readLines("/proc/cpuinfo")
	for _, line := range lines {
		if strings.HasPrefix(line, "cpu MHz") {
			idx := strings.Index(line, ":")
			if idx >= 0 {
				v, _ := strconv.ParseFloat(strings.TrimSpace(line[idx+1:]), 64)
				return v, ts
			}
		}
	}
	return 0, ts
}

func getNumCPU() int {
	// Count cpu directories in sysfs
	matches, _ := filepath.Glob("/sys/devices/system/cpu/cpu[0-9]*")
	if len(matches) > 0 {
		return len(matches)
	}
	return 1
}

func getCPUModel() string {
	lines, _ := exporter.readLines("/proc/cpuinfo")
	for _, line := range lines {
		if strings.HasPrefix(line, "model name") {
			idx := strings.Index(line, ":")
			if idx >= 0 {
				return strings.TrimSpace(line[idx+1:])
			}
		}
	}
	return "unknown"
}

func getKernelInfo() string {
	content, _ := exporter.readFile("/proc/version")
	return content
}

// getCPUCache reads cache info from sysfs.
// Returns map with keys like "L1d", "L1i", "L2", "L3" and values in bytes.
func getCPUCache() map[string]int64 {
	result := make(map[string]int64)
	seen := make(map[string]map[string]bool) // key -> shared_cpu_map -> seen

	pattern := "/sys/devices/system/cpu/cpu*/cache/index*"
	matches, _ := filepath.Glob(pattern)

	for _, dir := range matches {
		level, _ := exporter.readFile(filepath.Join(dir, "level"))
		ctype, _ := exporter.readFile(filepath.Join(dir, "type"))
		sizeStr, _ := exporter.readFile(filepath.Join(dir, "size"))
		cpuMap, _ := exporter.readFile(filepath.Join(dir, "shared_cpu_map"))

		if level == "" || sizeStr == "" {
			continue
		}

		// Build cache key (L1d, L1i, L2, L3)
		suffix := ""
		switch ctype {
		case "Data":
			suffix = "d"
		case "Instruction":
			suffix = "i"
		}
		key := "L" + level + suffix

		// Parse size (e.g., "32K", "256K", "8192K")
		size := parseSize(sizeStr)

		// Deduplicate by shared_cpu_map
		if seen[key] == nil {
			seen[key] = make(map[string]bool)
		}
		if !seen[key][cpuMap] {
			seen[key][cpuMap] = true
			result[key] += size
		}
	}
	return result
}

func parseSize(s string) int64 {
	s = strings.TrimSpace(s)
	mult := int64(1)
	if strings.HasSuffix(s, "K") {
		mult = 1024
		s = s[:len(s)-1]
	} else if strings.HasSuffix(s, "M") {
		mult = 1024 * 1024
		s = s[:len(s)-1]
	}
	v, _ := strconv.ParseInt(s, 10, 64)
	return v * mult
}

func nowMilli() int64 {
	return time.Now().UnixMilli()
}
